src/benchmark_runner.cc

```diff
@@ -221,6 +221,7 @@ BenchTimeType ParseBenchMinTime(const std::string& value) {
 
 BenchmarkRunner::BenchmarkRunner(
     const benchmark::internal::BenchmarkInstance& b_,
+    const std::shared_ptr<PerfCountersMeasurement>& pcm_,
```

Comment:
i'm still uncomfortable with the use of shared_ptr for this. we should be able to cleanly describe the lifetime and ownership of the PerfCountersMeasurement object, and shared_ptr elides that.

Reply:
Absolutely. You can see in RunBenchmark (benchmark.cc:351) that the object is created and then passed to N BenchmarkRunner objects (line 360). That is the reason why in the previous code we had a reference counter. So basically we are going from a hand-made intrusive pointer implementation to a proper STL shared_ptr, not much of a change. 

We could pass the pointer raw but there is an off-chance that in the future one of those objects linger beyond the scope of the counter and creates a segfault that would be hard to pinpoint. I think that's the main reason why we use smart pointers in the first place. But it is your code, I can definitely remove it - although I would not advise doing so.

Reply:
we know when all the runner objects complete, so that should be the scope of the unique_ptr.. the runners should get weak pointers.

i understand the ease that comes with shared pointers (and i agree the existing ref counting was worse than using them) but i would like to try to stick to something that requires us to understand the correctness of the code.

Reply:
If the runners get weak pointers then they will have to check the pointer every time they access it, which is an extra step. Im fine with that. But just so we are in the same page, weak pointers - as in [std::weak_ptr](https://en.cppreference.com/w/cpp/memory/weak_ptr) - have to be created from existing std::shared_ptr, they cannot be created from std::unique_ptr. We would have to pass a raw pointer from a unique_ptr.

Reply:
Why even give weak_ptr, since it;'d be a bug if access was attempted outside the scope of the unique_ptr, a naked ptr would suffice (and also avoid extra checks)

Reply:
yes sorry, i meant raw/naked pointer.  i still had shared_ptr in my head.

Reply:
That's okay with me. I just think that we are going back to the old C style with cowboy pointers while we have modern features to prevent a crash. But this is your code, I will make the change.

Reply:
@dmah42 Alright the commit is in! Please have a look.

Reply:
Out of complete curiosity, what is the general concern with shared pointers? I dont use them either but I use boost::intrusive_ptr quite extensively.

Reply:
to ensure that lifetimes are well-understood through the design of the code.  shared_ptr allows you to get away with thinking less about lifetimes and ownership, trusting that  things will just work out.  we prefer to avoid that _unless_ there are performance benefits and the object is immutable (to limit some classes of bugs).